Add discovery support to mqtt climate component.

dreizehnelf opened the pull request:
## Description:
The MQTT climate control did not support MQTT discovery - I needed it for a project and enabling it was very easy with minimal code changes. So maybe this is something that can be merged back to the main home-assistant codebase.

```

## Checklist:
  - [X ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass** (tox currently not working in my setup, sorry)

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) - Will update documentation there and submit PR as well.

dreizehnelf committed:
Add discovery support to mqtt climate component.

homeassistant commented:
Hello @dreizehnelf,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`df3ef10a56c64186f2be4dffb589461476d51603`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;15085&#x2F;commits&#x2F;df3ef10a56c64186f2be4dffb589461476d51603) - This commit has something that looks like an email address (michael@hengherr.de). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

dreizehnelf committed:
- Fix flake8 error (./homeassistant/components/climate/mqtt.py:130:1: D202 No blank lines allowed after function docstring)
- Fix test error (since climate component was expected not to work - changed it to "lock" component, which also does not have MQTT discovery support yet)

balloob commented on tests/components/mqtt/test_discovery.py:
```diff
@@ -55,7 +55,7 @@ def test_only_valid_components(mock_load_platform, hass, mqtt_mock, caplog):
     yield from async_start(hass, 'homeassistant', {})
 
-    async_fire_mqtt_message(hass, 'homeassistant/climate/bla/config', '{}')
+    async_fire_mqtt_message(hass, 'homeassistant/lock/bla/config', '{}')
```
Please read the rest of the test and fix it accordingly.

dreizehnelf replied:
Yeah... sorry. Need to fix up tox to be able to run the tests locally.

dreizehnelf replied:
Turns out - although the docs @ https://www.home-assistant.io/docs/mqtt/discovery/ don't mention locks as being supported, they are in the source code :) Sorry for missing that. I've setup tox locally and will track things down before recommitting.

dreizehnelf replied:
Hi Paulus - the test still fail, but I don't have enough background knowledge on how the whole setup with Travis works to judge whether this is a temporary thing or sth. that might be related to my commits. If there is anything I can do to fix the problems or move things further along, please let me know.

dreizehnelf committed:
Fix old assert statement to reflect new lock component usage

dreizehnelf committed:
Change invalid MQTT discovery component type from 'lock' to 'timer', since contrary to the documentation the lock component is properly supported when using MQTT discovery.

dreizehnelf committed:
Make configuration of invalid MQTT config component a single point of entry to prevent missing the assertion later in the code when changing.

dreizehnelf committed:
Add new testcases to cover not-yet-covered code paths in https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/climate/mqtt.py

dreizehnelf commented:
I'm really not sure what to do to get coveralls to sign off on the coverage - the details of the coveralls report doesn't really make sense. It list roughly 30 files that "have changed" (which they haven't) and claims coverage has decreased (which I can't understand, since I've actually increased coverage by adding test cases for code paths that had not been covered before). It might be related to coveralls sometimes getting confused (i.e. https://github.com/lemurheavy/coveralls-public/issues/1019) - but I've got no clue on how to fix it. Do all checks have to pass to get a PR merged into the main repository or are those just guidelines when reviewing? Any pointers highly appreciated!
